### PR TITLE
[model] fix: Gemma4 Dense fixes

### LINKED ISSUE #4610: [bug] Gemma4 Dense model lacks final logit softcapping
### Problem

In the [request](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4148) final logit softcapping was added only for MoE models

### Minimal repro

```shell
Run Gemma4
```

### Expected behavior

Logits are capped

### Affected area

area:model

### Regression?

Yes

### Environment

_No response_

### Logs

```shell

```

### PR DESCRIPTION
**Pull request by pavelgein**
# What does this PR do ?

Fix Gemma4 Dense model

# Changelog

1) Attach logit softcapping for dense models
2) Set `final_logit_softcapping` and `window_size` attributes. 3) Use TEDotProductAttention.

This class supports interleaving window attention, deducing attention type by `window_attn_skip_freq` config attribute

For small model, there is a parity check

```
======================================================================
  Parity [TEXT]: Megatron GPTModel (text)              vs  HF Gemma4ForCausalLM
  seq=16  batch=1  vocab=262144
  max |diff|  : 1.369141  (atol=3.0)
  mean |diff| : 0.090719
  worst token positions: [3, 12, 4] (diffs: ['1.3691', '0.6250', '0.5938'])
  --> PASSED
======================================================================
```

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)
Closes #4610

### CODE REVIEW

**Comment by cuichenx** [comment]
Hi, I believe #4645 addresses the issue. I also discovered a few other issues in the gemma 4 family and they're being addressed in #4683

**Comment by pavelgein** [comment]
Hi, what is the plan here? 
As far as I understand, there are three significantly overlapping PR, would be glad to help

**Comment by cuichenx** [comment]
> Hi, what is the plan here? As far as I understand, there are three significantly overlapping PR, would be glad to help

We're trying to get this merged to 0.5.1 release so we'll open the PRs to get it merged faster. I was trying to do comprehensive testing of all four gemma 4 models. Could you review 4645 and 4683? Let me know if anything is not addressed in those two PRs

**Comment by pavelgein** [comment]
#4645 was merged, and I think it lacks patch to parity check, since the latter does softcapping (so, outputs will be softcapped twice)

#4683 looks good to me, beside that there should be a broken [test](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/tests/functional_tests/test_groups/models/gemma/test_gemma4_conversion.py#L197), since Gemma does not support pipeline parallelism

**Comment by pavelgein** [comment]
Close, as everything has been done in other requests
